### PR TITLE
#2617 - Fixed issue with NPE in Generic Lists

### PR DESCRIPTION
**Pull request by davidjgonzalez**

### CODE REVIEW

**Review by badvision — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com)
 - #2581 - Versioned ClientLibs no longer works with proxied clientlibs
 - #2562 - Fixed cache refresh on versioned clientlibs request when enforceMd5 is false (default).
 - #2590- Fixed issue on 6.4.x with Service User mappings not being registered due to unsupported filename format of OSGi config.
```
Comment (HitmanInWis) [inline CHANGELOG.md:21]: Unrelated to this ticket, but might be good to fix the spacing issue here - `#2590 -`

File: bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java [hunk 2]
```diff
@@ -63,6 +63,11 @@ public AbstractResourceImpl(String path, String resourceType, String resourceSup
             meta.put("sling:resourceType", resourceType);
         }
         meta.put(JCR_PRIMARYTYPE, "nt:unstructured");
+        // Required property; Set this to the resource's path
+        meta.put(ResourceMetadata.RESOLUTION_PATH, this.path);
+        // Required property; This information does not exist in this context so set to pl
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java:68]: set to `pl`?  what is "pl"?

File: bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java [hunk 3]
```diff
@@ -63,6 +63,11 @@ public AbstractResourceImpl(String path, String resourceType, String resourceSup
...
+        meta.put(ResourceMetadata.RESOLUTION_PATH_INFO, "");
```
Comment (HitmanInWis) [inline bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java:69]: Since `metadata` can be passed into this function, should we check that `meta.get(ResourceMetadata.RESOLUTION_PATH_INFO)` is `null` prior to defaulting it here?

File: CHANGELOG.md (hunk 1 above)
Comment (davidjgonzalez) [inline CHANGELOG.md:21]:
```suggestion
- #2590 - Fixed issue on 6.4.x with Service User mappings not being registered due to unsupported filename format of OSGi config. 
```

File: bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java (hunk 2 above)
Comment (davidjgonzalez) [reply bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java:68]:
```suggestion
        // Required property; This information does not exist in this context so set to the abstract  resource's specified path
```

File: bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java [hunk 4]
```diff
@@ -63,6 +63,11 @@ public AbstractResourceImpl(String path, String resourceType, String resourceSup
...
+
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java:70]:
```suggestion
        // Required property; Set this to the resource's path
        meta.put(ResourceMetadata.RESOLUTION_PATH, this.path);
        
        if (meta.get(ResourceMetadata.RESOLUTION_PATH_INFO) == null) {
           // Required property; This information does not exist in this context so set to pl
          meta.put(ResourceMetadata.RESOLUTION_PATH_INFO, "");
        }

```

**Comment by davidjgonzalez** [comment]
@HitmanInWis good calls - that should address your feedback.

**Review by HitmanInWis — APPROVED** [review]
Code changes approved but I have not tested personally.

File: bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java [hunk 5]
```diff
@@ -63,6 +63,14 @@ public AbstractResourceImpl(String path, String resourceType, String resourceSup
             meta.put("sling:resourceType", resourceType);
         }
         meta.put(JCR_PRIMARYTYPE, "nt:unstructured");
+        // Required property; Set this to the resource's path
+        meta.put(ResourceMetadata.RESOLUTION_PATH, this.path);
+        
+        if (meta.get(ResourceMetadata.RESOLUTION_PATH_INFO) == null) {
+           // Required property; This information does not exist in this context so set to pl
```
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java:70]:
```suggestion
           // Required property; This information does not exist in this context so set to blank
```

**Comment by davidjgonzalez** [comment]
@kwin  any idea why this PR is failing in GH Actions? Tests pass on my local. Same with.. https://github.com/Adobe-Consulting-Services/acs-aem-commons/pull/2626

**Comment by davidjgonzalez** [comment]
@badvision  any idea why this test if failing .. it looks like the problem is that the mocking of `rr.resolve(..)` isnt working, and thus the `.resolve(..)` returns null.


```
when(rr.resolve("/content/pageA")).thenReturn(pageA);
when(rr.resolve("/content/pageB")).thenReturn(pageB);
when(rr.resolve("/content/pageC")).thenReturn(new NonExistingResource(rr, "/content/pageC"));
when(rr.resolve("/content/pageD")).thenReturn(new NonExistingResource(rr, "/content/pageD"));
when(rr.resolve("/content/pageE")).thenReturn(new NonExistingResource(rr, "/content/pageE"));

Map<String, Object> values = new HashMap<>();
values.put("sourcePath", "/content");
values.put("propertyRegex", "^/(etc|content)/.+");
values.put("excludeProperties", "ignoredRef");

ProcessInstanceImpl instance = new ProcessInstanceImpl(getControlledProcessManager(), tool, "broken references");
instance.init(rr, values);
instance.run(rr);

Map<String, EnumMap<Report, Object>> reportData = tool.getReportData();
assertEquals(3, reportData.size());
```

I have to imagine that in ...

`...filter(path -> ResourceUtil.isNonExistingResource(resource.getResourceResolver().resolve(path)))`

the `resource.getResourceResolver()...` Isnt returning the mocked `rr`  but its hard to follow the code to figure out where everything is going. Oddly this does work on my local macOS on JDK 8 and 11 ... just breaks in Github Actions...

**Comment by davidjgonzalez** [comment]
Re-opening. Originally closed to incorporate kwin's suggestion using SyntheticResource, but that changes the classes API quite a bit, and cascades through to a bunch of other uses...  i'd rather just put this quick fix in now, and tackle a larger refactoring later.

**Comment by davidjgonzalez** [comment]
This ended up being quite a big bigger than anticipated, as it turns out the initial implementation conflated ResourceMetadata with a resource's properties.

File: bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java [hunk 6]
```diff
@@ -63,6 +63,14 @@ public AbstractResourceImpl(String path, String resourceType, String resourceSup
             meta.put("sling:resourceType", resourceType);
         }
         meta.put(JCR_PRIMARYTYPE, "nt:unstructured");
```
[the remark this answers is not in the document]
Comment (davidjgonzalez) [reply bundle/src/main/java/com/adobe/acs/commons/mcp/form/AbstractResourceImpl.java]: Sticking w this AbstractResourceImpl for this PR - we can adjust to Synthetic later. Making the switch seems like a non-trivial change due to some of the impl decisions on this feature.
